content/2020-05-22-book-review-writing-an-interpreter-in-go/index.md

```diff
@@ -0,0 +1,581 @@
+
+#### keywords
+
+These are reserved words like "function" and "let" in our language. Similar to how "mut" is reserved in Rust, but not JavaScript.
+
+#### repl means something
+
+I remember a few years ago hearing about [repl.it](https://repl.it/) and thinking to myself, "What a strange name!" It was only then in 1.5 of this book where Thorsten enlightened me. It stands for "read evaluate print loop."
+
+> Python has a REPL, Ruby has one, JavaScript runtime has one, most Lisps have one and a lot of other languages too.
+>
+> -Ball 33
+
+It reads your input, lets the interpreter evaluate it, prints the result or output and then repeats. Another aha moment from this book.
+
+### Chapter 2 - Parsing
+
+The parser is the "software component that takes input data (frequently text) and builds a data structure" (Ball 36).
+
+> "Code is data, data is code" is something you hear a lot from Lisp programmers.
+>
+> -Ball 37
+
+#### serialization languages
+
+I don't think I had heard this term/concept mentioned before, but he was referring to JSON, YAML, TOML, INI, etc.
+
+#### syntactic analysis
+
+Again, I don't think I had heard this before.
+
+> ...the process of parsing is also call syntactic analysis
+>
+> -Ball 39
+
+#### parser generator
+
+Another new term, Thorsten explains that parser generators are "tools that, when fed with a formal description of a language, produce parsers as their output" (Ball 39).
+
+To be honest, this is still a bit foreign to me. Of course when I google it, the term "[compiler-compiler](https://en.wikipedia.org/wiki/Compiler-compiler)" comes up which doesn't help. I think I'll leave it at that for now, but may revisit.
+
+> Parsing is one of he most well-understood branches of computer science...
+>
+> -Ball 40
+
+#### top-down vs bottom-up vs recursive decent parsing
+
+I bet I could write a blog post on this alone (I don't know all of the differences, and won't explore them here). The method we take in the book is recursive decent parsing which he mentions is a "top down operator precedence" parser and also called "Pratt parser" after Vaughan Pratt.
+
+#### binding vs identifier vs expression
+
+I wrote a note to myself "define binding identifier vs. expression." Let's give it a shot. Let statements like `let x = 5;` bind the expression "5" to the identifier "x". And as Thorsten reminds us, "Expressions produce values, statments don't" (Ball 42). To be clear, if we take this `let x = 5;` it is a statement because it does not produce a value. However, this `5` does produce a value. It's like a tupperware container with some food inside. The tupperware doesn't give us food, but if we go inside and look, we get food.
```

Comment:
There's an `e` missing in `statement`:

```suggestion
I wrote a note to myself "define binding identifier vs. expression." Let's give it a shot. Let statements like `let x = 5;` bind the expression "5" to the identifier "x". And as Thorsten reminds us, "Expressions produce values, statements don't" (Ball 42). To be clear, if we take this `let x = 5;` it is a statement because it does not produce a value. However, this `5` does produce a value. It's like a tupperware container with some food inside. The tupperware doesn't give us food, but if we go inside and look, we get food.
```